File: src/utils/timing.ts
export class Stopwatch {
  private startTime: number = 0;
  private endTime: number = 0;

  start(): void {
    this.startTime = performance.now();
  }

  stop(): number {
    this.endTime = performance.now();
    return this.elapsed();
  }

  elapsed(): number {
    if (this.endTime === 0) {
      return performance.now() - this.startTime;
    }
    return this.endTime - this.startTime;
  }

  reset(): void {
    this.startTime = 0;
    this.endTime = 0;
  }
}

export function calculatePercentiles(
  values: number[], 
  percentiles: number[]
): Record<string, number> {
  if (values.length === 0) {
    return Object.fromEntries(percentiles.map(p => [`p${p}`, 0]));
  }

  const sorted = [...values].sort((a, b) => a - b);
  const result: Record<string, number> = {};

  for (const p of percentiles) {
    const index = (p / 100) * (sorted.length - 1);
    const lower = Math.floor(index);
    const upper = Math.ceil(index);
    const weight = index % 1;

    if (lower === upper) {
      result[`p${p}`] = sorted[lower];
    } else {
      result[`p${p}`] = sorted[lower] * (1 - weight) + sorted[upper] * weight;
    }
  }

  return result;
}

export function timeout<T>(promise: Promise<T>, ms: number): Promise<T> {
  return Promise.race([
    promise,
    new Promise<never>((_, reject) => 
      setTimeout(() => reject(new Error(`Timeout after ${ms}ms`)), ms)
    )
  ]);
}

File: jarvis/utils/timing.py
"""Timing helpers, ported from src/utils/timing.ts."""
from __future__ import annotations

import asyncio
import time
from typing import Awaitable, TypeVar

T = TypeVar("T")


class Stopwatch:
    def __init__(self) -> None:
        self._start = 0.0
        self._end = 0.0

    def start(self) -> None:
        self._start = time.perf_counter() * 1000

    def stop(self) -> float:
        self._end = time.perf_counter() * 1000
        return self.elapsed()

    def elapsed(self) -> float:
        if self._end == 0:
            return time.perf_counter() * 1000 - self._start
        return self._end - self._start

    def reset(self) -> None:
        self._start = 0.0
        self._end = 0.0


def calculate_percentiles(values: list[float], percentiles: list[int]) -> dict[str, float]:
    if not values:
        return {f"p{p}": 0.0 for p in percentiles}

    sorted_values = sorted(values)
    result: dict[str, float] = {}

    for p in percentiles:
        index = (p / 100) * (len(sorted_values) - 1)
        lower, upper = int(index // 1), -(-int(index) // 1)
        lower = int(index)
        upper = min(lower + 1, len(sorted_values) - 1)
        weight = index - lower

        if lower == upper:
            result[f"p{p}"] = sorted_values[lower]
        else:
            result[f"p{p}"] = sorted_values[lower] * (1 - weight) + sorted_values[upper] * weight

    return result


async def with_timeout(coro: Awaitable[T], ms: float) -> T:
    try:
        return await asyncio.wait_for(coro, timeout=ms / 1000)
    except asyncio.TimeoutError as exc:
        raise TimeoutError(f"Timeout after {ms}ms") from exc
